sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,458 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_RUNTIME_EPMP_H_
+#define OPENTITAN_SW_DEVICE_LIB_RUNTIME_EPMP_H_
+
+#include <stdint.h>
+
+/**
+ * RISC-V Enhanced Physical Memory Protection (EPMP).
+ *
+ * Specifications:
+ *  - PMP Enhancements for memory access and execution prevention on Machine
+ *    mode
+ *    https://docs.google.com/document/d/1Mh_aiHYxemL0umN3GTTw8vsbmzHZ_nxZXgjgOUzbvc8
+ *  - RISC-V Privileged Specfication (v. 20190608)
+ *    https://github.com/riscv/riscv-isa-manual/releases/download/Ratified-IMFDQC-and-Priv-v1.11/riscv-privileged-20190608.pdf
+ *  - Ibex PMP
+ *    https://ibex-core.readthedocs.io/en/latest/03_reference/pmp.html
+ *
+ * Assumptions (should be initialized in assembly but can be verified using
+ * `epmp_check`):
+ *   * Machine Mode Whitelist Policy is enabled (mseccfg.MMWP = 1)
+ *   * Machine Mode Lockdown is disabled (mseccfg.MML = 0)
+ */
+
+/**
+ * EPMP constants.
+ *
+ * TODO: there will probably be more of these.
+ */
+typedef enum epmp_constants {
+  kEpmpGranularity = 0,
+  kEpmpNumRegions = 16,
+} epmp_constants_t;
+
+/**
+ * Toggle for Rule Locking Bypass (mseccfg.RLB).
+ */
+typedef enum epmp_rlb_toggle {
+  kEpmpRlbToggleDisabled = 0, /**< mseccfg.RLB = 0 */
+  kEpmpRlbToggleEnabled,      /**< mseccfg.RLB = 1 */
+} epmp_rlb_toggle_t;
+
+/**
+ * EPMP state.
+ *
+ * A copy of a possible EPMP control register state.
+ */
+typedef struct epmp_state {
+  uint8_t pmpcfg[kEpmpNumRegions];
+  uintptr_t pmpaddr[kEpmpNumRegions];
```

Comment:
Can you add comments describing these member variables?

Reply:
Done